### DISCUSSION TITLE
antonyms in languages other than German and English

### DISCUSSION THREAD

**Discussion by chlor** [goodmami/wn#161] [Q&A] [answered]
How to get antonyms from language sources other than German or English language.
Here is my way to get antonyms:

`                if synset.get_related('antonym'):
                    antonyms.update(list(itertools.chain(*[ant.lemmas() for ant in synset.get_related('antonym')])))
                for sense in synset.senses():
                    if sense.get_related('antonym'):
                        ant = sense.get_related('antonym')
                        for a in ant:
                            antonyms.update(a.word().lemma())
`

**Comment by chlor** [comment]
I 've tested it with Italian, French an Spanish and found no examples.

**Comment by fcbond** [comment]
The wordnets from omw 1.* do not have their own relations but piggyback on
PWN.  So lemmas in synsets in any language that have antonyms in PWN should
be antonyms, or at least have opposite meanings.  You can get a bigger list
if you also expand through similar-to, ...

**Comment by chlor** [comment]
Thanks for the quick reply, how can I read antonyms from OWN networks? The code fragments I mentioned are not sufficient.

**Comment by fcbond** [comment]
Hi,

Not optimized at all, but this should work.  It gets the antonym relations from English and then looks them up in Japanese.


```python
import wn
en = wn.Wordnet('omw-en:1.4')
ja = wn.Wordnet('omw-ja:1.4')

### a set of tuples of ili ids of antonyms,
### sorted internally (so we only get one way: id1 < id2)
antonyms = set()

### store ilis for antonym pairs
for synset in en.synsets():
    ### check synsets (there aren't any)
    if syn_anto := synset.get_related('antonym'):
        print(synset, syn_anto)
    ### check senses
    for sense in synset.senses():
        if sen_anto := sense.get_related('antonym'):
            for anto in sen_anto:
                antonyms.add(tuple(sorted([sense.synset().ili.id,
                                           anto.synset().ili.id])))
### print for Japanese
for (a1, a2) in antonyms:
    s1 = ja.synsets(ili=a1)
    s2 = ja.synsets(ili=a2)
    print(s1,s2)
```

**Comment by chlor** [comment, marked as answer]
Thanks, it works!
Here is my extension to get antoynms:

````import itertools

import wn
en = wn.Wordnet('omw-en:1.4')
ja = wn.Wordnet('omw-ja:1.4')

### a set of tuples of ili ids of antonyms,
### sorted internally (so we only get one way: id1 < id2)
antonyms = set()
antonyms_txt = set()

### store ilis for antonym pairs
for synset in en.synsets():
    if syn_anto := synset.get_related('antonym'):
        print(synset, syn_anto)
    for sense in synset.senses():
        if sen_anto := sense.get_related('antonym'):
            for anto in sen_anto:
                antonyms.add(tuple(sorted([sense.synset().ili.id, anto.synset().ili.id])))

dict_ant = {}

for (a1, a2) in antonyms:
    s1 = ja.synsets(ili=a1)
    s2 = ja.synsets(ili=a2)
    for s in s1:
        for lem in s.lemmas():
            got_ants = [s.lemmas() for s in s2]
            if got_ants:
                dict_ant[lem] = got_ants

for d in dict_ant:
    print(d, *itertools.chain.from_iterable(dict_ant[d]))

print(len(dict_ant))
````
